File: android/app/src/main/java/com/dibay/app/nativevideo/NativeVideoCallUiPresenter.java
package com.dibay.app.nativevideo;

import android.content.Context;
import com.dibay.app.R;

/** Maps Native Video Runtime session/state to render-only UI model. */
public final class NativeVideoCallUiPresenter {
  public enum Phase {
    INCOMING,
    DIALING,
    CONNECTING,
    CONNECTED,
    ENDING
  }

  public static final class Model {
    public final Phase phase;
    public final String peerName;
    public final String statusText;
    public final String avatarInitial;
    public final boolean showIncomingActions;
    public final boolean showActiveActions;
    public final boolean showConnectedControls;
    public final boolean showVideoSurfaces;
    public final boolean showLocalPreview;
    public final boolean showStatusOverlay;
    public final boolean showDuration;
    public final String endButtonLabel;
    public final String cameraLabel;

    Model(
        Phase phase,
        String peerName,
        String statusText,
        String avatarInitial,
        boolean showIncomingActions,
        boolean showActiveActions,
        boolean showConnectedControls,
        boolean showVideoSurfaces,
        boolean showLocalPreview,
        boolean showStatusOverlay,
        boolean showDuration,
        String endButtonLabel,
        String cameraLabel) {
      this.phase = phase;
      this.peerName = peerName;
      this.statusText = statusText;
      this.avatarInitial = avatarInitial;
      this.showIncomingActions = showIncomingActions;
      this.showActiveActions = showActiveActions;
      this.showConnectedControls = showConnectedControls;
      this.showVideoSurfaces = showVideoSurfaces;
      this.showLocalPreview = showLocalPreview;
      this.showStatusOverlay = showStatusOverlay;
      this.showDuration = showDuration;
      this.endButtonLabel = endButtonLabel;
      this.cameraLabel = cameraLabel;
    }
  }

  private NativeVideoCallUiPresenter() {}

  public static Model build(Context context, NativeVideoCallRuntime.Session session, NativeVideoCallRuntime.State state) {
    Context app = context != null ? context.getApplicationContext() : null;
    String peerName = resolvePeerName(session);
    Phase phase = resolvePhase(session, state);
    String statusText = resolveStatusText(app, phase);
    String avatarInitial = initialFromName(peerName);
    boolean ending = phase == Phase.ENDING;
    boolean incoming = phase == Phase.INCOMING;
    boolean connected = phase == Phase.CONNECTED;
    boolean dialingOrConnecting = phase == Phase.DIALING || phase == Phase.CONNECTING;
    boolean videoPhase = phase == Phase.CONNECTING || phase == Phase.CONNECTED;
    String endLabel =
        phase == Phase.DIALING
            ? safeString(app, R.string.dibay_video_call_cancel)
            : safeString(app, R.string.dibay_video_call_end);
    String cameraLabel = safeString(app, R.string.dibay_video_camera_on);
    return new Model(
        phase,
        peerName,
        statusText,
        avatarInitial,
        incoming && !ending,
        (dialingOrConnecting || connected) && !ending,
        connected && !ending,
        videoPhase && !ending,
        videoPhase && !ending,
        !videoPhase && !ending,
        connected && !ending,
        endLabel,
        cameraLabel);
  }

  public static Phase resolvePhase(NativeVideoCallRuntime.Session session, NativeVideoCallRuntime.State state) {
    if (state == NativeVideoCallRuntime.State.ENDING
        || state == NativeVideoCallRuntime.State.ENDED
        || state == NativeVideoCallRuntime.State.FAILED) {
      return Phase.ENDING;
    }
    if (state == NativeVideoCallRuntime.State.CONNECTED) {
      return Phase.CONNECTED;
    }
    if (state == NativeVideoCallRuntime.State.RINGING) {
      return Phase.INCOMING;
    }
    if (session != null && session.initiator && state == NativeVideoCallRuntime.State.CONNECTING) {
      return Phase.DIALING;
    }
    if (state == NativeVideoCallRuntime.State.ACCEPTING || state == NativeVideoCallRuntime.State.CONNECTING) {
      return Phase.CONNECTING;
    }
    return Phase.CONNECTING;
  }

  private static String resolvePeerName(NativeVideoCallRuntime.Session session) {
    if (session == null || session.callerName == null || session.callerName.trim().isEmpty()) {
      return "DIBAY";
    }
    return session.callerName.trim();
  }

  private static String resolveStatusText(Context app, Phase phase) {
    if (app == null) return "";
    switch (phase) {
      case INCOMING:
        return safeString(app, R.string.dibay_video_call_incoming);
      case DIALING:
        return safeString(app, R.string.dibay_video_call_calling);
      case CONNECTING:
        return safeString(app, R.string.dibay_video_call_connecting);
      case CONNECTED:
        return safeString(app, R.string.dibay_video_call_connected);
      case ENDING:
      default:
        return safeString(app, R.string.dibay_video_call_ending);
    }
  }

  private static String initialFromName(String name) {
    if (name == null || name.isEmpty()) return "D";
    return name.substring(0, 1).toUpperCase();
  }

  private static String safeString(Context app, int resId) {
    try {
      return app.getString(resId);
    } catch (RuntimeException error) {
      return "";
    }
  }
}

File: ios/App/App/Call/Video/NativeVideoCallUiPresenter.swift
import Foundation

/** Maps Native Video Runtime session/state to render-only UI model (Android parity). */
enum NativeVideoCallUiPresenter {
  enum Phase: Equatable, Sendable {
    case incoming
    case dialing
    case connecting
    case connected
    case ending
  }

  struct Model: Equatable, Sendable {
    let phase: Phase
    let peerName: String
    let statusText: String
    let avatarInitial: String
    let showIncomingActions: Bool
    let showActiveActions: Bool
    let showConnectedControls: Bool
    let showVideoSurfaces: Bool
    let showLocalPreview: Bool
    let showStatusOverlay: Bool
    let showDuration: Bool
    let endButtonLabel: String
    let cameraLabel: String
  }

  static func build(session: NativeVideoCallSession?, state: NativeVideoCallRuntimeState) -> Model {
    let peerName = resolvePeerName(session)
    let phase = resolvePhase(session: session, state: state)
    let statusText = resolveStatusText(phase: phase)
    let avatarInitial = initialFromName(peerName)
    let ending = phase == .ending
    let incoming = phase == .incoming
    let connected = phase == .connected
    let dialingOrConnecting = phase == .dialing || phase == .connecting
    let videoPhase = phase == .connecting || phase == .connected
    let endLabel = phase == .dialing ? Copy.cancel : Copy.end
    return Model(
      phase: phase,
      peerName: peerName,
      statusText: statusText,
      avatarInitial: avatarInitial,
      showIncomingActions: incoming && !ending,
      showActiveActions: (dialingOrConnecting || connected) && !ending,
      showConnectedControls: connected && !ending,
      showVideoSurfaces: videoPhase && !ending,
      showLocalPreview: videoPhase && !ending,
      showStatusOverlay: !videoPhase && !ending,
      showDuration: connected && !ending,
      endButtonLabel: endLabel,
      cameraLabel: Copy.cameraOn
    )
  }

  static func resolvePhase(
    session: NativeVideoCallSession?,
    state: NativeVideoCallRuntimeState
  ) -> Phase {
    switch state {
    case .ending, .ended, .failed:
      return .ending
    case .connected:
      return .connected
    case .ringing:
      return .incoming
    case .accepting, .connecting:
      if let session, session.initiator, state == .connecting {
        return .dialing
      }
      return .connecting
    }
  }

  private static func resolvePeerName(_ session: NativeVideoCallSession?) -> String {
    let name = session?.callerName.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    return name.isEmpty ? "DIBAY" : name
  }

  private static func resolveStatusText(phase: Phase) -> String {
    switch phase {
    case .incoming:
      return Copy.incoming
    case .dialing, .connecting:
      return Copy.connecting
    case .connected:
      return Copy.connected
    case .ending:
      return Copy.ending
    }
  }

  private static func initialFromName(_ name: String) -> String {
    guard let first = name.first else { return "D" }
    return String(first).uppercased()
  }

  private enum Copy {
    static let incoming = "영상 통화 수신"
    static let connecting = "연결 중…"
    static let connected = "통화 중"
    static let ending = "통화 종료 중"
    static let end = "종료"
    static let cancel = "취소"
    static let cameraOn = "카메라 켬"
    static let cameraOff = "카메라 끔"
  }

  static func cameraLabel(enabled: Bool) -> String {
    enabled ? Copy.cameraOn : Copy.cameraOff
  }
}
